### PR TITLE
Change healing by sleeping to be affected by sanity, darkness (or blindfold), and earmuffs.

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request

Depending on the mob's sanity level, it can have a positive or negative boost to healing effects while sleeping.  Sleeping in darkness, wearing a blindfold, and using earmuffs also counts as a healing bonus.  Beauty sleep is very important for 2D spessmen.

## Why It's Good For The Game

This is a small gameplay change that rewards players for keeping their sanity at good levels.  Also depression has also been linked with impeding wound healing in real life.  The placebo effect on peoples minds is strenuously documented and I think it would be cool to see it in the game.

## Changelog

:cl:
expansion: Healing by sleeping is now affected by sanity, sleeping in darkness (or using a blindfold), and using earmuffs.  The healing from sleeping in a bed was slightly decreased.
/:cl:

### CODE REVIEW

**Comment by JinGiant** [comment]
could you please point to the singulo and where that is in real life?

**Comment by timothymtorres** [comment]
Black holes exist in real life fam sorry to break your immersion.

**Comment by JinGiant** [comment]
> Black holes exist in real life fam sorry to break your immersion.

Do they also form and destroy the world when you delam the supermatter in real life

**Comment by catrock31** [comment]
> Do they also form and destroy the world when you delam the supermatter in real life
yeah fam sorry to break your immersion.

**Comment by JinGiant** [comment]
> > Do they also form and destroy the world when you delam the supermatter in real life
> > yeah fam sorry to break your immersion.

Give me one example

**Comment by ATH1909** [comment]
does this mean that anaesthetic will damage insane people until they hit 80% health?

**Comment by bruhlookatthisdood** [comment]
the supermatter doesn't exist in real life either, guess we should remove it.

**Comment by TheBonded** [comment]
why are people so mad about this change? tell me one genuine scenario when you are actively utilizing the slow heal when sleeping in a bed with a blanket or the slow heal while sleeping with morphine where this will actively impede you due to low mood. you can't. because these aren't even used ever.

**Comment by ilikesaying** [comment]
islep pr

the graytider trying not to make any noise as the captain is sleeping

**Comment by JinGiant** [comment]
> the supermatter doesn't exist in real life either, guess we should remove it.

you purposefully missed the point.

**Comment by timothymtorres** [comment]
All this talk about realism put me in the mood to make it even more detailed.  The healing rate is now also affected by being in the darkness (alternatively using a blindfold) and using earmuffs.  Everyone knows how critical important beauty sleep is to have robust 2D spessmen running the station.

**Comment by Rhials** [comment]
> > the supermatter doesn't exist in real life either, guess we should remove it.
> 
> you purposefully missed the point.

Then can you give an actual point for people to take away from your replies? So far all I can gather is that you don't like this PR because it's... too realistic?

**Comment by bruhlookatthisdood** [comment]
so..
when is this being merged?

File: code/datums/status_effects/debuffs.dm [hunk 1]
```diff
@@ -180,31 +180,69 @@
...
 	if(owner.maxHealth)
 		var/health_ratio = owner.health / owner.maxHealth
-		var/healing = -0.2
+		var/healing = HEALING_SLEEP_DEFAULT
+
+		// having high spirits helps us recover
+		var/datum/component/mood/mood = owner.GetComponent(/datum/component/mood)
+		if(mood != null)
+			switch(mood.sanity_level)
+				if(SANITY_LEVEL_GREAT)
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs.dm:194]: Use these defines in the code that defined their values please

File: code/datums/status_effects/debuffs.dm [hunk 2]
```diff
@@ -180,31 +180,69 @@
...
+					healing = 0
+				if(SANITY_LEVEL_UNSTABLE)
+					healing = 0
+				if(SANITY_LEVEL_CRAZY)
+					healing = 0.1
+				if(SANITY_LEVEL_INSANE)
+					healing = 0.2
+
+		var/turf/rest_turf = get_turf(owner)
+		var/is_sleeping_in_darkness = rest_turf.get_lumcount() <= 0.2 // switch 0.2 to the LIGHTING_IS_DARK define when Paper DLC gets merged
```
Comment (LemonInTheDark) [inline code/datums/status_effects/debuffs.dm:208]: why the fuck is the paper dlc pr adding light level defines? reading?

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
cute

File: code/datums/status_effects/debuffs.dm (hunk 2 above)
Comment (timothymtorres) [reply code/datums/status_effects/debuffs.dm:208]: Exactly.  The plan is to later on make it something that is usable by the `canUseTopic` proc as a bitflag. (so people can't do stuff like surgery in the dark, etc.)

File: code/datums/status_effects/debuffs.dm (hunk 1 above)
Comment (timothymtorres) [reply code/datums/status_effects/debuffs.dm:194]: Fixed with https://github.com/tgstation/tgstation/pull/65713/commits/77316bbf8606685e60d098046dfe3d5dce54a8c8

**Comment by JinGiant** [comment]
> > > the supermatter doesn't exist in real life either, guess we should remove it.
> > 
> > 
> > you purposefully missed the point.
> 
> Then can you give an actual point for people to take away from your replies? So far all I can gather is that you don't like this PR because it's... too realistic?

Yes

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 3]
```diff
@@ -178,31 +178,69 @@
...
 			healing -= 0.1
 			break //Only count the first bedsheet
-		if(health_ratio > 0.8)
-			owner.adjustBruteLoss(healing)
-			owner.adjustFireLoss(healing)
-			owner.adjustToxLoss(healing * 0.5, TRUE, TRUE)
-		owner.adjustStaminaLoss(healing)
+
+		if(healing < 0)
+			if(health_ratio > 0.8)
```
Comment (ExcessiveUseOfCobblestone) [inline code/datums/status_effects/debuffs/debuffs.dm:230]: &&

Comment (timothymtorres) [reply code/datums/status_effects/debuffs/debuffs.dm]: Applied with https://github.com/tgstation/tgstation/pull/65713/commits/17ade4f6970d08b8c9479d3384079a3de07c7bf2

File: code/datums/status_effects/debuffs.dm (hunk 2 above)
Comment (timothymtorres) [reply code/datums/status_effects/debuffs.dm]: - #66308 Was merged so I can finally use the proper define with https://github.com/tgstation/tgstation/pull/65713/commits/415b8408e3bcb35c40098d5925adef711236e8d3

**Comment by LemonInTheDark** [comment]
@ExcessiveUseOfCobblestone are you happy with this?

**Comment by moocowswag** [comment]
Make it so sleeping with a plushie gives you a mood buff?

File: code/datums/status_effects/debuffs/debuffs.dm [hunk 4]
```diff
@@ -178,31 +178,68 @@
 	ADD_TRAIT(owner, TRAIT_KNOCKEDOUT, TRAIT_STATUS_EFFECT(id))
 	tick_interval = initial(tick_interval)
 
+#define HEALING_SLEEP_DEFAULT -0.2
+
 /datum/status_effect/incapacitating/sleeping/tick()
 	if(owner.maxHealth)
 		var/health_ratio = owner.health / owner.maxHealth
-		var/healing = -0.2
+		var/healing = HEALING_SLEEP_DEFAULT
```
Comment (Ryll-Ryll) [inline code/datums/status_effects/debuffs/debuffs.dm:186]: Flip the signs of all of this, `healing` should not be a negative number if you're regaining health

Comment (timothymtorres) [reply code/datums/status_effects/debuffs/debuffs.dm:186]:
```dm
owner.adjustBruteLoss(healing)
owner.adjustFireLoss(healing)
```
These procs are how healing is used.  Adding a positive number is damage and negative number is healing damage.

Comment (tralezab) [reply code/datums/status_effects/debuffs/debuffs.dm:171]: healing can be positive and then you do -healing on the proc itself!

Comment (timothymtorres) [reply code/datums/status_effects/debuffs/debuffs.dm:171]: Applied with 112d67996c357d2d0c1f824ef1140154b6cc403f

**Comment by timothymtorres** [comment]
@ExcessiveUseOfCobblestone this is ready unless there is something else you want me to change.

**Pull request merged**